source/fontloader.ts

```diff
@@ -0,0 +1,227 @@
+        const file = stripped(pairs.get('file')!, ['"', '\r']);
+
+        const pngPath: string = path + file.split('.')[0] + '.png';
+
+        return fontFace.glyphTexture.load(pngPath);
+    }
+
+    protected handleChar(stream: Array<string>, fontFace: FontFace): void {
+        const pairs = this.readKeyValuePairs(stream,
+            ['id', 'x', 'y', 'width', 'height', 'xoffset', 'yoffset', 'xadvance']);
+
+        const index: number = parseInt(pairs.get('id')!, 10);
+        assert(index > 0, `handleChar: glyph index is not greater than 0.`);
+
+        const glyph = new Glyph();
+
+        glyph.index = index;
+
+        const extentScale: GLfloat2 = [
+            1.0 / fontFace.glyphTextureExtent[0],
+            1.0 / fontFace.glyphTextureExtent[1],
+        ];
+        const extent: GLfloat2 = [
+            parseFloat(pairs.get('width')!),
+            parseFloat(pairs.get('height')!),
+        ];
+
+        glyph.subTextureOrigin = [
+            parseFloat(pairs.get('x')!) * extentScale[0],
+            1.0 - (parseFloat(pairs.get('y')!) + extent[1]) * extentScale[1],
+        ];
+
+        glyph.extent = extent;
+
+        glyph.subTextureExtent[0] = extent[0] * extentScale[0];
+        glyph.subTextureExtent[1] = extent[1] * extentScale[1];
+
+        glyph.setBearing(fontFace.base,
+            parseFloat(pairs.get('xoffset')!),
+            parseFloat(pairs.get('yoffset')!),
+        );
+
+        glyph.advance = parseFloat(pairs.get('xadvance')!);
+
+        fontFace.addGlyph(glyph);
+    }
+
+    protected handleKerning(stream: Array<string>, fontFace: FontFace): void {
+        const pairs = this.readKeyValuePairs(stream, ['first', 'second', 'amount']);
+
+        const first: number = parseInt(pairs.get('first')!, 10);
+        assert(first > 0, `handleKerning: first is not greater than 0.`);
+
+        const second: number = parseInt(pairs.get('second')!, 10);
+        assert(second > 0, `handleKerning: second is not greater than 0.`);
+
+        const kerning: number = parseFloat(pairs.get('amount')!);
+
+        fontFace.setKerning(first, second, kerning);
+    }
+
+    protected readKeyValuePairs(stream: Array<string>, mandatoryKeys: Array<string>): StringPairs {
+
+        let key: string;
+        let value: string;
+
+        const pairs: StringPairs = new Map<string, string>();
+
+        for (const s of stream) {
+            const pair = s.split('=');
+            key = pair[0];
+            value = pair[1];
+            pairs.set(key, value);
+        }
+
+        // check if all required keys are provided
+        let valid = true;
+        for (const mandatoryKey of mandatoryKeys) {
+            valid = valid && pairs.has(mandatoryKey);
+        }
+
+        if (!valid) {
+            console.warn('Not all required keys are provided! Mandatory keys: ' + mandatoryKeys);
+            return new Map<string, string>(); // typescript does not allow writing `new StringPairs()`
```

Comment:
wouldn't it be better to then return at least the found keys, instead of undefined or an empty Map?
Or: error instead of warning.
I'd prefer the latter. What's your opinion on that?

Reply:
returning undefined means that I get errors "Object is possibly undefined". To avoid those errors, I'd clutter the code with undefined-checks. So I think I still prefer returning an empty Map.
(plus: warning --> error)